Byte-update lowering: fixup invariant for array-comprehension

tautschnig opened the pull request:
In case of an array comprehension the number of operands must not be
used to sanity check the update offset (any update offset would be
valid). This was observed on several device driver benchmarks in
SV-COMP. One example (with --unwind 2):
ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--wireless--rndis_wlan.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.i

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Byte-update lowering: fixup invariant for array-comprehension

In case of an array comprehension the number of operands must not be
used to sanity check the update offset (any update offset would be
valid). This was observed on several device driver benchmarks in
SV-COMP. One example (with --unwind 2):
ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--wireless--rndis_wlan.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.i

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 43ec36dcc09848cea6f938ee611379e19d2cff6f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113350664

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -1819,7 +1819,8 @@ static exprt lower_byte_update_struct(
     {
       offset = from_integer(0, src.offset().type());
       INVARIANT(
-        value_as_byte_array.operands().size() > -*offset_bytes,
```
Would the byte array have a type reflecting its constant size? If so how about using the type-size which would work for comprehensions too?

tautschnig replied:
If it's an array comprehension then it would almost certainly be of non-constant size, and therefore couldn't be evaluated at this time.

tautschnig committed:
Honour update bound when updating struct members

We must not unconditionally use bytes extracted from the update value
when the size of the update value is not constant: the bytes extracted
may be out-of-bounds and thus non-deterministic.

tautschnig commented:
This now includes a test, which in turn prompted me to check upon another case, which resulted in a slightly bigger bugfix. All in here now with tests.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -1860,20 +1860,40 @@ static exprt lower_byte_update_struct(
+    optionalt<exprt> member_update_bound;
+    if(non_const_update_bound.has_value())
+    {
+      const exprt remaining_update_bytes = typecast_exprt::conditional_cast(
```
This construction is pretty difficult to follow; I'd appreciate a comment with pseudocode

tautschnig replied:
Comments added and code partly re-arranged to make things a bit easier to follow.

smowton reviewed (approved):
Just a request for a comment

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d6a13bda7c7972fadbf6cf049660f6258bf0b6b9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113727071

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 8cc423c0f58e9306d035eed86c375da389510da1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113956031
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
